File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.*;

class Solution {
    public int solution(String[] friends, String[] gifts) {
        // 총 친구들의 수 n
        int n = friends.length;
        // 총 선물 받은 결과 리스트 result
        int[] result = new int [n];
        
        // 친구들의 index를 담을 map
        Map<String, Integer> friendMap = new HashMap<>();
        
        for (int i = 0; i < n; i++) {
            // 이름, index로 map에 추가 
            friendMap.put(friends[i], i);  
        }
        
        // 선물 지수 리스트
        int[] count = new int[n];
        // 선물 주고받은 내용 담을 map
        int[][] map = new int[n][n];
        
        // gift를 돌며 선물 표와 지수 count
        for (String gift: gifts) {
            // 선물 준 사람 이름 gName, index gIndex
            String gName = gift.split(" ")[0];
            int gIndex = friendMap.get(gName);
            
            // 선물 받은 사람 이름 rName, index rIndex
            String rName = gift.split(" ")[1];
            int rIndex = friendMap.get(rName);
            
            // 선물 지수 count
            count[gIndex]++;
            count[rIndex]--;
            
            // 선물 주고 받은 표 count
            map[gIndex][rIndex]++;
        }
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                // 자기 자신에게 선물 줄 수 없음
                if (i >= j) continue;
                
                // 1. 선물 기록이 없거나 선물 개수가 같은 경우
                if (map[i][j] == map[j][i]) {
                    // 선물 지수가 같은 경우
                    if (count[i] == count[j]) continue;
                    
                    // 선물 지수가 다른 경우
                    if (count[i] > count[j])
                        result[i]++;
                    else 
                        result[j]++;
                    
                    continue;
                }
                
                // 2. 선물 기록이 있고 개수가 다른 경우
                if (map[i][j] > map[j][i])
                    result[i]++;
                else 
                    result[j]++;
            }
        }
        
        // 가장 선물 많은 받을 친구 찾고 선물 수 return
        int maxCount = 0;
        for (int k : result) {
            // 선물 더 많이 받을 친구면 갱신 
            maxCount = maxCount < k ? k : maxCount;
        }
        
        return maxCount;
    }
}

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.js
function solution(friends, gifts) {
    // 전체 사람 수
    let n = friends.length;
    
    // 각 사람 별 인덱스 번호 저장
    let map = new Map();
    
    for(let i = 0; i < n; i++) {
        map.set(friends[i], i);
    }
    
    // 선물 주고 받은 내역 2차원 배열로 저장
    let history = new Array(n).fill().map(() => new Array(n).fill(0));
    
    for (gift of gifts) {
        // 주는 사람의 인덱스 sIdx, 받는 사람의 인덱스 rIdx
        const sIdx = map.get(gift.split(' ')[0]);
        const rIdx = map.get(gift.split(' ')[1]);
        
        // 선물 주고 받은 2차원 배열
        history[sIdx][rIdx]++;
    }
    
    console.log(history);
    
    // 총 선물을 받을 개수를 담을 배열
    const answer = new Array(n).fill(0);
    
    // 선물 주고 받은 기록 비교
    for (let j = 0; j < n; j++) {
        for (let k = j + 1; k < n; k++) {
            // 1. 주고받은 선물 개수가 같은 경우
            if (history[j][k] === history[k][j]) {
                // 선물 지수 계산
                let a = 0; let b = 0;
                
                let rowSum = 0; let colSum = 0;
                
                // 열, 행의 합 계산
                for (let num = 0; num < n; num++) {
                    colSum += history[num][j]; // 받은 선물 개수
                    rowSum += history[j][num]; // 준 선물 개수
                }
                
                a = rowSum - colSum;
                
                rowSum = 0; colSum = 0;
                
                for (let num = 0; num < n; num++) {
                    colSum += history[num][k]; // 받은 선물 개수
                    rowSum += history[k][num]; // 준 선물 개수
                }
                
                b = rowSum - colSum;
                
                if (a > b) {
                    answer[j]++;
                } else if (b > a) {
                    answer[k]++;
                }
            }
            
            // 2. 한명이 더 많이 선물은 준 경우
            else if (history[j][k] > history[k][j]) {
                answer[j]++;
            } else {
                answer[k]++;
            }
        }
    }
    
    let maxValue = 0;
    
    for(value of answer) {
        maxValue = maxValue < value ? value : maxValue;
    }
    
    return maxValue;
}